### ISSUE TITLE
BUG: pd.factorize with read-only datetime64 numpy array raises ValueError

### ISSUE THREAD

**Issue by ruchirgarg05** [pandas-dev/pandas#35650]
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python

In [1]: pandas.__version__
[PYFLYBY] import pandas
Out[1]: u'0.24.2'

In [2]: arr = numpy.array([numpy.datetime64('2015-11-20T15:06:58.000')])

In [3]: arr.dtype
Out[3]: dtype('<M8[ms]')

In [4]: arr.flags.writeable = False

[PYFLYBY] import pandas as pd
In [5]: pd.factorize(arr)
```

#### Problem description

[Construction with non-mutable datetime64 strings]

#### Expected Output
(array([0]), array(['2015-11-20T15:06:58.000000000'], dtype='datetime64[ns]'))

#### Output of ``pd.show_versions()``

<details>
pandas/_libs/tslibs/conversion.pyx in pandas._libs.tslibs.conversion.ensure_datetime64ns()

/usr/local/python/python-2.7/std/lib/python2.7/site-packages/pandas/_libs/tslibs/conversion.so in View.MemoryView.memoryview_cwrapper()

/usr/local/python/python-2.7/std/lib/python2.7/site-packages/pandas/_libs/tslibs/conversion.so in View.MemoryView.memoryview.__cinit__()

ValueError: buffer source array is read-only
</details>

**Comment by jreback** [comment]
pls try on 1.1 and/or master

**Comment by simonjayhawkins** [comment]
works on pandas latest (and master) fails on v1.0.5

```
>>> pd.__version__
'1.1.0'
>>>
>>> arr = np.array([np.datetime64("2015-11-20T15:06:58.000")])
>>>
>>> arr.dtype
dtype('<M8[ms]')
>>>
>>> arr.flags.writeable = False
>>>
>>> pd.factorize(arr)
(array([0], dtype=int64), array(['2015-11-20T15:06:58.000000000'], dtype='datetime64[ns]'))
>>>
```

**Comment by simonjayhawkins** [comment]
> works on pandas latest (and master) fails on v1.0.5

This was fixed by #34844.

a8bbe75dc2d6e05900648490a96ec8bb7b9e8e0e is the first new commit
commit a8bbe75dc2d6e05900648490a96ec8bb7b9e8e0e
Author: Joris Van den Bossche <jorisvandenbossche@gmail.com>
Date:   Thu Jun 18 08:47:44 2020 +0200

    BUG: fix construction from read-only non-ns datetime64 numpy array (#34844)

**Comment by avinashpancham** [comment]
take
